Support for BigInteger values in Table

radeusgd opened the pull request:
### Pull Request Description

- Fixes #7354
- And also closes #7712 
- Refactors how we handle numeric ops - ensuring that the 'kernels' are placed all in one place and selected based on storage types.

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

radeusgd committed:
add tests for big integers

radeusgd committed:
fixing some mistakes in added tests

radeusgd committed:
add a check

radeusgd committed:
check for big integer scalar argument in arithmetic operations in-memory

radeusgd committed:
add a test

radeusgd committed:
BigInteger builder - easier to detect BigIntegers if they have their own builder, but then also made sense to create BigInteger storage. for now no arith ops - so warning is attached

TODO: casting

radeusgd committed:
casts

jdunkerley reviewed (approved):
This looks a good first step so we don't get the weird Mixed behaviour.

radeusgd committed:
testing a few approaches

radeusgd committed:
Refactoring operations to more common form, adding ops for BigInteger

radeusgd committed:
WIP: comparisons

I think both of my changes don't work on Mixed types well enough yet

radeusgd committed:
remove code disallowing decimals pt.1

radeusgd committed:
remove code disallowing decimals pt.2

radeusgd committed:
refactor, impl IsIn for BigInteger

radeusgd committed:
fixing Mixed numeric comparisons, various fixes

radeusgd committed:
fixing Mixed numeric comparisons pt2

radeusgd committed:
Test arithmetic, remove warnings

radeusgd committed:
Remove obsolete UnsupportedFeature

radeusgd committed:
more tests

radeusgd committed:
fixing tests
fixed common type logic for more numerics
edge cases with builders

radeusgd committed:
javafmt

radeusgd committed:
more tests!

radeusgd committed:
add type check to Column.at

radeusgd committed:
fix a typo in test

radeusgd committed:
improve common type logic for min/max/fill_nothing, etc.

radeusgd committed:
gather warnings from fillMissing

radeusgd committed:
javafmt

GregoryTravis commented on distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type_Helpers.enso:
```diff
@@ -17,20 +17,32 @@ reconcile_types current new = case current of
+            if (precision == new_precision) && (scale == new_scale) then new else
+                # TODO at some point we may want a more clever merging of precision and scale, for now we don't use them too much anyway so we just default to Nothing if they do not agree
+                Value_Type.Decimal
+        Value_Type.Integer _ -> Value_Type.Decimal precision scale
+        Value_Type.Byte      -> Value_Type.Decimal precision scale
```
```suggestion
        Value_Type.Integer _ -> current
        Value_Type.Byte      -> current
```

radeusgd committed:
add a BigInt roundtrip test

radeusgd committed:
Fix uploading BigIntegers to DB

radeusgd committed:
add test for is_* ops for a biginteger column

radeusgd committed:
Update handling of Decimal column upload.

Added tests covering the scenarios in Postgres, including handling of very large values that do not fit in the 1000 digits limit.

Tests are passing, modulo checking for warnings (TODO)

radeusgd committed:
Add checks for Inexact_Type_Coercion

radeusgd committed:
Implement basic warnings for inexact types

radeusgd committed:
Add a polyglot test

GregoryTravis commented on distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Enso_Types.enso:
```diff
@@ -22,13 +23,16 @@ most_specific_value_type value use_smallest=False =
+                        value_type = Storage.to_value_type storage_type
+                        # We do a small rewrite here - for integers we always return the Integer type, even if the value is small enough to fit in a Byte.
+                        if value_type == Value_Type.Byte then Value_Type.Integer Bits.Bits_16 else value_type
+                True -> Value_Type.Decimal precision=Nothing scale=0
```
Ah, now I understand better why `Value_Type.Decimal` is being used for `BigInteger`.

GregoryTravis commented on std-bits/table/src/main/java/org/enso/table/data/column/operation/map/numeric/arithmetic/NumericBinaryOpImplementation.java:
```diff
@@ -0,0 +1,269 @@
+ */
+public abstract class NumericBinaryOpImplementation<T extends Number, I extends Storage<? super T>> extends BinaryMapOperation<T, I> implements NumericBinaryOpDefinition {
+
+  // Regardless of input type, our integer operations return 64-bit integers.
```
What about the `BigInteger` case?

radeusgd replied:
Right! This comment got outdated, thanks for noticing, I will update it.

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/7354-table-bigint-warning

radeusgd committed:
javafmt

radeusgd committed:
remove unused files

radeusgd committed:
CR comment

radeusgd committed:
add a test

jdunkerley commented on distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Column_Fetcher.enso:
```diff
@@ -83,6 +84,18 @@ long_fetcher bits =
 
+## PRIVATE
+biginteger_fetcher : Column_Fetcher
+biginteger_fetcher =
```
nit `big_integer_fetcher`.

jdunkerley commented on test/Tests/polyglot-sources/enso-test-java-helpers/src/main/java/org/enso/table_test_helpers/PolyglotHelpers.java:
```diff
@@ -0,0 +1,16 @@
+public class PolyglotHelpers {
+  /* Creates a big integer that is larger than long can fit. */
+  public static BigInteger createBigBigIntegerComingFromJava() {
+    return BigInteger.TWO.pow(70);
```
is it worth making this no something that can be stored in a double exactly?
```suggestion
    return BigInteger.TWO.pow(70).minus(1);
```

radeusgd replied:
Actually, `2^70` is (in a way) _not_ exactly representable.

I was surprised, but that seems to be the case:
```
> 2^70
>>> 1180591620717411303424
> (2^70)+0.0
>>> 1.1805916207174113E21
> (2^70)+0.0 . truncate
>>> 1180591620717411300000
```

At least the roundtrip back to integer is inexact - that is because the conversion from `double` to `BigDecimal` happens at base 10 and not 2.

radeusgd replied:
(But I still have changed the test in question)

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/7354-table-bigint-warning

radeusgd committed:
CR2

radeusgd committed:
CR3, adding some tests
